### ISSUE TITLE
Need a Spark.awaitStopped() method

### ISSUE THREAD

**Issue by kdubezerra** [perwendel/spark#705]
The `Spark.stop()` method stops the server, setting the internal `initialized` variable to `false`. The problem is that this is all done asynchronously, in a separate thread (in the `Spark.Service` class):

```
/**
 * Stops the Spark server and clears all routes
 */
public synchronized void stop() {
    new Thread(() -> {
        if (server != null) {
            routes.clear();
            server.extinguish();
            latch = new CountDownLatch(1);
        }

        staticFilesConfiguration.clear();
        initialized = false;
    }).start();
}
```

However, I found no way of checking whether the service has actually stopped. This is problematic because Spark creates a singleton service, so shutting down a server and restarting it without actually finishing the server process may lead to problems. In a more concrete use case, this creates problems when running several unit tests, where each unit test starts a server, runs the test, and shuts the server down.

This issue would be solved by providing a method like `Spark.awaitStopped()` that returned only when the thread started in `Spark.Service.stop()` had finished.

**Comment by jakaarl** [comment]
I did some quick experimenting, shouldn't be too hard to implement.

**Comment by jakaarl** [comment]
My quick hack of an implementation basically does a `join()` on the thread if waiting is required. Sound okay (@perwendel @tipsy)? Need to write tests at least before raising a PR.

**Comment by kdubezerra** [comment]
I think a better way to go would be with a condition variable with a lock on the `initialized` field... At the end of the `stop` thread, you'd signal the condvar. The awaitStopped() would get the lock, check the value of `initialized`, waiting on the condvar if necessary, then releasing the lock.

**Comment by jakaarl** [comment]
@kdubezerra sounds more complicated than necessary, what benefits would the added complexity bring?

**Comment by kdubezerra** [comment]
@jakaarl That's a good question. To me, it looks like a textbook example of when a condition variable should be used: all it depends on is the `initialized` variable being false (hence the condition). And, fwiw, using the condvar will consist of ~5 (10 at most) lines of code. But maybe using a `Thread.join` would be just as correct and probably simpler.

I have another question though: what if you call `Spark.awaitStopped()` before starting the server? What if it's called after the Thread spawned by `Spark.stop()` has already finished? Do you keep a reference to that thread just in case someone calls `awaitStopped()`?. My 2 cents is that the condvar would cover all these cases, but I'm just as happy with whatever implementation you guys provide :)

**Comment by jakaarl** [comment]
My implementation of `awaitStopped()` actually stops and waits for it to finish, instead of waiting on a separate call to `stop()` to finish. Very simple, but doesn't quite do what was asked or what the name implies. I'll take a new look at this when time allows. But would a blocking version of `stop()` that I implemented be desirable feature as well?

**Comment by kdubezerra** [comment]
Yeah, that would definitely help! I actually remembered of guava services... As another suggestion, one can extend `AbstractService` and implement `doStop()` (the blocking version) as well as `stopAsync()` (similar to Spark's current one). The `awaitRunning()` and `awaitTerminated()` methods would come for free.

**Comment by mddubey** [comment]
Hey guys,
  I am having a similar issue. Here is my code for it

port(Integer.parseInt(mci_port));
`try {
    fhirPatientValidator = new FhirPatientValidator(mciProperties);
} catch (Exception e) {
    logger.error("Unable to create FHIR Patient Validator",e);
    Spark.stop();
}`

Actually, I am trying to construct an object, if that fails I want the server to be stopped and do nothing. But I figured out  it is still going and executing lines after that. I was thinking of putting `System.exit(statusCode)`, but not sure whether that will be a good way of stopping it or not.

Thanks in advance.

**Comment by perwendel** [comment]
Fixed in #730
